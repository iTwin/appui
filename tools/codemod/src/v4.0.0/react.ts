/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
import type { API, FileInfo, Options } from "jscodeshift";
import { useExtensions } from "../utils/Extensions";

export default function transformer(
  file: FileInfo,
  api: API,
  options: Options
) {
  const j = api.jscodeshift;
  useExtensions(j);

  const root = j(file.source);

  root.rename(
    "@itwin/appui-react:CommonWidgetProps",
    "@itwin/appui-react:Widget"
  );
  root.rename("@itwin/appui-react:WidgetProps", "@itwin/appui-react:Widget");

  return root.toSource(options.printOptions);
}
