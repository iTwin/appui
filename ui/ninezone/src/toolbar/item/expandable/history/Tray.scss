@import "@bentley/ui-core/lib/colors";
@import "../../../../_geometry";

.nz-toolbar-item-expandable-history-tray {
  $border-radius: 3px;
  $extend-tool-history-tab-width: 5px;
  $triangle-width: 3px;
  $triangle-color: $text-light;

  display: flex;

  > .nz-items {
    display: flex;

    & > * {
      display: none;
    }

    & > *:first-child {
      display: block;
    }
  }

  &.nz-is-extended {
    > .nz-items {
      & > * {
        display: block;
      }
    }
  }

  > .nz-tab {
    display: none;
    width: $extend-tool-history-tab-width;
    background: $background-4;
    align-items: center;
    justify-content: center;
  }

  &.nz-direction-left {
    flex-direction: row-reverse;

    > .nz-items {
      flex-direction: row-reverse;

      & > *:last-child {
        border-top-left-radius: $border-radius;
        border-bottom-left-radius: $border-radius;
      }
    }

    > .nz-tab {
      border-radius: 0;
      border-top-left-radius: $border-radius;
      border-bottom-left-radius: $border-radius;

      > .nz-triangle {
        @include triangle-left($width: $triangle-width, $color: $triangle-color);
      }
    }
  }

  &.nz-direction-top {
    flex-direction: column-reverse;

    > .nz-items {
      flex-direction: column-reverse;

      & > *:last-child {
        border-radius: 0;
        border-top-left-radius: $border-radius;
        border-top-right-radius: $border-radius;
      }
    }

    > .nz-tab {
      width: auto;
      height: $extend-tool-history-tab-width;
      border-radius: 0;
      border-top-left-radius: $border-radius;
      border-top-right-radius: $border-radius;

      > .nz-triangle {
        @include triangle-top($width: $triangle-width, $color: $triangle-color);
      }
    }
  }

  &.nz-direction-right {
    flex-direction: row;

    > .nz-items {
      flex-direction: row;

      & > *:last-child {
        border-radius: 0;
        border-top-right-radius: $border-radius;
        border-bottom-right-radius: $border-radius;
      }
    }

    > .nz-tab {
      border-radius: 0;
      border-top-right-radius: $border-radius;
      border-bottom-right-radius: $border-radius;

      > .nz-triangle {
        @include triangle-right($width: $triangle-width, $color: $triangle-color);
      }
    }
  }

  &.nz-direction-bottom {
    flex-direction: column;

    > .nz-items {
      flex-direction: column;

      & > *:last-child {
        border-radius: 0;
        border-bottom-left-radius: $border-radius;
        border-bottom-right-radius: $border-radius;
      }
    }

    > .nz-tab {
      width: auto;
      height: $extend-tool-history-tab-width;
      border-radius: 0;
      border-bottom-left-radius: $border-radius;
      border-bottom-right-radius: $border-radius;

      > .nz-triangle {
        @include triangle-bottom($width: $triangle-width, $color: $triangle-color);
      }
    }
  }

  &.nz-tab-is-visible {
    > .nz-tab {
      display: flex;
    }

    > .nz-items {
      & > *:last-child {
        border-radius: 0;
      }
    }
  }
}
