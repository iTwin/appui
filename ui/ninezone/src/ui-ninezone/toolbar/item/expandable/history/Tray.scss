/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/

@import "@bentley/ui-core/lib/ui-core/style/themecolors";
@import "../../../../_geometry";

.nz-toolbar-item-expandable-history-tray {
  $border-radius: 3px;
  $extend-tool-history-tab-width: 5px;
  $triangle-width: 3px;
  $triangle-color: $buic-icon-color;

  display: flex;
  pointer-events: auto;

  > .nz-items {
    display: flex;

    > * {
      display: none;
    }

    > *:first-child {
      display: block;
    }
  }

  &.nz-is-extended {
    > .nz-items {
      > * {
        display: block;
      }
    }
  }

  > .nz-extend-indicator {
    display: none;
    width: $extend-tool-history-tab-width;
    background: $buic-background-4;
    align-items: center;
    justify-content: center;

    &:after {
      content: '';
    }
  }

  &.nz-direction-left {
    flex-direction: row-reverse;

    > .nz-items {
      flex-direction: row-reverse;

      > *:last-child {
        border-top-left-radius: $border-radius;
        border-bottom-left-radius: $border-radius;
      }
    }

    > .nz-extend-indicator {
      border-radius: 0;
      border-top-left-radius: $border-radius;
      border-bottom-left-radius: $border-radius;

      &:after {
        @include triangle-left($width: $triangle-width, $color: $triangle-color);
      }
    }
  }

  &.nz-direction-top {
    flex-direction: column-reverse;

    > .nz-items {
      flex-direction: column-reverse;

      > *:last-child {
        border-radius: 0;
        border-top-left-radius: $border-radius;
        border-top-right-radius: $border-radius;
      }
    }

    > .nz-extend-indicator {
      width: auto;
      height: $extend-tool-history-tab-width;
      border-radius: 0;
      border-top-left-radius: $border-radius;
      border-top-right-radius: $border-radius;

      &:after {
        @include triangle-top($width: $triangle-width, $color: $triangle-color);
      }
    }
  }

  &.nz-direction-right {
    flex-direction: row;

    > .nz-items {
      flex-direction: row;

      > *:last-child {
        border-radius: 0;
        border-top-right-radius: $border-radius;
        border-bottom-right-radius: $border-radius;
      }
    }

    > .nz-extend-indicator {
      border-radius: 0;
      border-top-right-radius: $border-radius;
      border-bottom-right-radius: $border-radius;

      &:after {
        @include triangle-right($width: $triangle-width, $color: $triangle-color);
      }
    }
  }

  &.nz-direction-bottom {
    flex-direction: column;

    > .nz-items {
      flex-direction: column;

      > *:last-child {
        border-radius: 0;
        border-bottom-left-radius: $border-radius;
        border-bottom-right-radius: $border-radius;
      }
    }

    > .nz-extend-indicator {
      width: auto;
      height: $extend-tool-history-tab-width;
      border-radius: 0;
      border-bottom-left-radius: $border-radius;
      border-bottom-right-radius: $border-radius;

      &:after {
        @include triangle-bottom($width: $triangle-width, $color: $triangle-color);
      }
    }
  }

  &.nz-extend-is-visible {
    > .nz-extend-indicator {
      display: flex;
    }

    > .nz-items {
      > *:last-child {
        border-radius: 0;
      }
    }
  }
}
