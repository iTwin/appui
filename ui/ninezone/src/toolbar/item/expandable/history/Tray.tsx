/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
/** @module Toolbar */

import * as classnames from "classnames";
import * as React from "react";

import Direction, { DirectionHelpers } from "../../../../utilities/Direction";
import CommonProps from "../../../../utilities/Props";

import "./Tray.scss";

export interface Entry<TItem> {
  key: string;
  item: TItem;
}

export interface HistoryProps<TItem> extends Array<Entry<TItem>> {
}

export const MAX_ITEM_CNT = 4;

export const addItem = <TItem extends {}>(key: string, item: TItem, props: HistoryProps<TItem>): HistoryProps<TItem> => {
  const itemToRemove = props.findIndex((entry) => {
    return entry.key === key;
  });
  if (itemToRemove > -1)
    props.splice(itemToRemove, 1);

  if (props.unshift({ key, item }) > MAX_ITEM_CNT)
    props.pop();

  return props;
};

export interface TrayProps extends CommonProps {
  direction?: Direction;
  isExpanded?: boolean;
  onIsExpandedChange?: (isExpanded: boolean) => void;
}

export default class Tray extends React.Component<TrayProps> {
  public render() {
    const count = React.Children.count(this.props.children);
    const tabVisible = count > 1 && !this.props.isExpanded;
    const className = classnames(
      "nz-toolbar-item-expandable-history-tray",
      DirectionHelpers.getCssClassName(this.props.direction || Direction.Left),
      tabVisible && "nz-tab-is-visible",
      this.props.isExpanded && "nz-is-extended",
      this.props.className);

    return (
      <div
        className={className}
        style={this.props.style}
        onMouseEnter={this.handleOnMouseEnter}
        onMouseLeave={this.handleOnMouseLeave}
      >
        <div className="nz-items">
          {this.props.children}
        </div>
        <div className="nz-tab">
          <div className="nz-triangle" />
        </div>
      </div>
    );
  }

  private handleOnMouseEnter = () => {
    this.props.onIsExpandedChange && this.props.onIsExpandedChange(true);
  }

  private handleOnMouseLeave = () => {
    this.props.onIsExpandedChange && this.props.onIsExpandedChange(false);
  }
}
