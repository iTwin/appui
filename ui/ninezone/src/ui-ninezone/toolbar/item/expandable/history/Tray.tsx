/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Toolbar
 */

import * as classnames from "classnames";
import * as React from "react";
import { CommonProps, NoChildrenProps, flattenChildren } from "@bentley/ui-core";
import { Direction, DirectionHelpers } from "../../../../utilities/Direction";
import "./Tray.scss";

// tslint:disable: deprecation

/** Key to identify history item.
 * @alpha
 * @deprecated History tray removed from design standard.
 */
export type HistoryKey = number | string;

/** A single entry in items history.
 * @alpha
 * @deprecated History tray removed from design standard.
 */
export interface HistoryEntry<TItem> {
  /** Entry key. */
  key: HistoryKey; // tslint:disable-line: deprecation
  /** Entry item. */
  item: TItem;
}

/** History defines actual structure of history items.
 * @alpha
 * @deprecated History tray removed from design standard.
 */
export type History<TItem> = Array<HistoryEntry<TItem>>; // tslint:disable-line: deprecation

/** Helper to manage history entries.
 * @alpha
 * @deprecated History tray removed from design standard.
 */
export class HistoryManager {
  public constructor(public readonly maxItemCount: number) {
  }

  /**
   * Adds specified item to history and returns a new state of history.
   * @note Immutable operation.
   */
  public addItem<TItem extends {}>(key: HistoryKey, item: TItem, history: History<TItem>): History<TItem> { // tslint:disable-line: deprecation
    const itemToRemove = history.findIndex((entry) => {
      return entry.key === key;
    });
    const newHistory = itemToRemove < 0 ? [...history] : [
      ...history.slice(0, itemToRemove),
      ...history.slice(itemToRemove + 1),
    ];

    if (newHistory.unshift({ key, item }) > this.maxItemCount)
      newHistory.pop();

    return newHistory;
  }
}

/** History manager as defined by 9-Zone UI specification.
 * @alpha
 * @deprecated History tray removed from design standard.
 */
// tslint:disable-next-line:variable-name
export const DefaultHistoryManager = new HistoryManager(4); // tslint:disable-line: deprecation

/** Properties of [[HistoryTray]] component.
 * @alpha
 * @deprecated History tray removed from design standard.
 */
export interface HistoryTrayProps extends CommonProps, NoChildrenProps {
  /** Extend direction of tray. */
  direction?: Direction;
  /** Describes if the tray is extended. If this is false, at most 1 item is visible. */
  isExtended?: boolean;
  /** Items of tray component. I.e. [[HistoryItem]], [[Icon]] */
  items?: React.ReactNode;
  /** Function called when history tray should be extended or shrank. */
  onIsHistoryExtendedChange?: (isExtended: boolean) => void;
}

/** History tray used in [[ExpandableItem]] component.
 * @alpha
 * @deprecated History tray removed from design standard.
 */
export class HistoryTray extends React.PureComponent<HistoryTrayProps> { // tslint:disable-line: deprecation
  public render() {
    const items = flattenChildren(this.props.items);
    const count = React.Children.count(items);
    const isExtendIndicatorVisible = count > 1 && !this.props.isExtended;
    const className = classnames(
      "nz-toolbar-item-expandable-history-tray",
      DirectionHelpers.getCssClassName(this.props.direction || Direction.Left),
      isExtendIndicatorVisible && "nz-extend-visible",
      this.props.isExtended && "nz-extended",
      this.props.className);

    return (
      <div
        className={className}
        style={this.props.style}
        onMouseEnter={this._handleMouseEnter}
        onMouseLeave={this._handleMouseLeave}
      >
        <div className="nz-items">
          {this.props.items}
        </div>
        <div className="nz-extend-indicator" />
      </div>
    );
  }

  private _handleMouseEnter = () => {
    this.props.onIsHistoryExtendedChange && this.props.onIsHistoryExtendedChange(true);
  }

  private _handleMouseLeave = () => {
    this.props.onIsHistoryExtendedChange && this.props.onIsHistoryExtendedChange(false);
  }
}
