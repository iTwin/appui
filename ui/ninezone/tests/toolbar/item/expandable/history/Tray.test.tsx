/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
import { mount, shallow } from "enzyme";
import * as React from "react";

import Tray, { MAX_ITEM_CNT, addItem } from "@src/toolbar/item/expandable/history/Tray";

describe("<Tray />", () => {
  it("should render", () => {
    mount(<Tray />);
  });

  it("renders correctly", () => {
    shallow(<Tray />).should.matchSnapshot();
  });

  it("renders expanded", () => {
    shallow(<Tray isExpanded />).should.matchSnapshot();
  });

  it("renders tabs", () => {
    shallow(<Tray><br /><br /></Tray>).should.matchSnapshot();
  });

  it("MAX_ITEM_CNT should eq 4", () => {
    MAX_ITEM_CNT.should.eq(4);
  });

  it("addItem should add history entry", () => {
    const sut = addItem("key1", {}, []);
    sut.length.should.eq(1);
  });

  it("addItem should add 4 items most", () => {
    let sut = addItem("key1", {}, []);
    sut = addItem("key2", {}, sut);
    sut = addItem("key3", {}, sut);
    sut = addItem("key4", {}, sut);
    sut = addItem("key5", {}, sut);
    sut.length.should.eq(4);
  });

  it("addItem should re-add item with same key", () => {
    let sut = addItem("key1", {}, []);
    sut = addItem("key2", {}, sut);
    sut = addItem("key1", {}, sut);
    sut.length.should.eq(2);
  });

  it("addItem should work as FIFO structure", () => {
    let sut = addItem("key1", {}, []);
    sut = addItem("key2", {}, sut);
    sut[1].key.should.eq("key1");
    sut[0].key.should.eq("key2");
  });
});
