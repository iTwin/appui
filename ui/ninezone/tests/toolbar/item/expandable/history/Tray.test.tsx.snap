// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tray /> renders correctly 1`] = `
<div
  className="nz-toolbar-item-expandable-history-tray nz-direction-left"
>
  <div
    className="nz-items"
  />
  <div
    className="nz-extend-indicator"
  />
</div>
`;

exports[`<Tray /> renders extended 1`] = `
<div
  className="nz-toolbar-item-expandable-history-tray nz-direction-left nz-is-extended"
>
  <div
    className="nz-items"
  />
  <div
    className="nz-extend-indicator"
  />
</div>
`;

exports[`<Tray /> renders items 1`] = `
<div
  className="nz-toolbar-item-expandable-history-tray nz-direction-left nz-extend-is-visible"
>
  <div
    className="nz-items"
  >
    <br />
    <br />
  </div>
  <div
    className="nz-extend-indicator"
  />
</div>
`;
